Update documentation to reflect RFC002 and RFC005 (#1237)
tweag/nickel: doc/manual/merging.md

- They are both of a primitive data type `Number`, `Bool`, `Enum`, `String` and
  they are equal
- They are both arrays and they are equal (checked by generating an application

>   of the lazy contract `contract.Equal`)

Reviewer: ```suggestion
  of the lazy contract `std.contract.Equal`)
```

## Merging records with metadata

Metadata can be attached to values thanks to the `|` operator. Metadata

> currently include contract annotation, default value, merge priority, and

Reviewer: I actually find the previous formulation "Metadata currently includes contract annotations" more natural 😕 I don't have a good rationale, though.

Author: I don't have a good one either :sweat_smile: I think data is technically plural (the plural of _datum_) but is both plural and singular in practice. Maybe should I summon a native speaker such as @matthew-healy

documentation. We describe in this section how metadata interacts with merging.

> Note that metadata can only be syntactically attached to record fields, at the

Reviewer: ```suggestion
Note that metadata can only be syntactically attached to record fields, with the
```

Author: Ah, good catch of this horrible Frenchism :stuck_out_tongue:

exception of type and contract annotations, which can appear anywhere in a
freestanding expression such as `(x | Num) + 1`. However, a contract annotation

> outside of a record field isn't considered a metadata (it's a mere contract

Reviewer: ```suggestion
outside of a record field isn't considered metadata (it's a mere contract
```

#### Specification

> To each field definition `foo = val` is associated a priority `p(val)`. When
> merging two common fields `value_left` and `value_right`, the results is either
> the one with the highest priority (that overrides the other), or the two are

Reviewer: ```suggestion
Each field definition `foo = val` is assigned a priority `p(val)`. When
merging two common fields `value_left` and `value_right`, the result is either
the one with the highest priority (that overrides the other), or the two are
```

including being dropped in favor of another value, the final value for `foo` has
to respect the contract as well or the evaluation will fail accordingly.

This is only true for *contracts attached directly to record fields* (either

> directly, on coming from an enclosing record contract). In particular, `{foo |

Reviewer: ```suggestion
directly, or coming from an enclosing record contract). In particular, `{foo |
```

> This is only true for *contracts attached directly to record fields* (either
> directly, or coming from an enclosing record contract). In particular, `{foo |
> Num = 1} & {foo | force = "bar"}` will fail, but `{foo = (1 | Num)} & {foo |

Reviewer 2: ```suggestion
This is only true for *contracts attached directly to record fields* (either
directly, or coming from an enclosing record contract). In particular, `{foo |
Number = 1} & {foo | force = "bar"}` will fail, but `{foo = (1 | Number)} & {foo |
```

> force = "bar"}` will succeed. In the later case, the contract is not considered
> to be a field metadata, but a local contract check, which is not propagated by
> merging.

Reviewer: ```suggestion
force = "bar"}` will succeed. In the latter case, the contract is not considered
to be field metadata, but a local contract check, which is not propagated by
merging.
```

`common: LeftType` that can just be considered as an additional contract
`Left0`.

The accumulated contracts are applied lazily: as long as the field's value isn't
requested, contracts are accumulated but not applied. This makes it possible to
build a value piecewise, whereas the intermediate values don't necessarily
respect the contract. For example:

```nickel
let FooContract = {
  required_field1,
  required_field2,
} in
{ foo | FooContract}
& { foo.required_field1 = "here" }
& { foo.required_field2 = "here" }
```

Running the above program succeeds, although the intermediate value
`{foo.required_field1 = "here"}` doesn't respect `FooContract` (it misses the
field `required_field2`).

> If we try to observe the intermediate result (`deep_seq` recrusively forces the

Reviewer: ```suggestion
If we try to observe the intermediate result (`deep_seq` recursively forces the
```
